onnxruntime/core/providers/cuda/nn/conv.cc

```diff
@@ -287,128 +291,181 @@ Status Conv<T, NHWC>::UpdateState(OpKernelContext* context, bool bias_expected)
+        auto bias_options = cudnn_frontend::graph::Pointwise_attributes().set_mode(cudnn_frontend::PointwiseMode_t::ADD);
+
+        s_.cudnn_fe_Y = s_.cudnn_fe_graph->pointwise(conv_output, s_.cudnn_fe_B, bias_options);
+      } else {
+        s_.cudnn_fe_Y = conv_output;
+      }
+      s_.cudnn_fe_Y->set_output(true);
+
+      CUDNN_FE_RETURN_IF_ERROR(s_.cudnn_fe_graph->validate());
+      auto handle = GetCudnnHandle(context);
+      CUDNN_FE_RETURN_IF_ERROR(s_.cudnn_fe_graph->build_operation_graph(handle));
 
-      cudnnConvolutionFwdAlgoPerf_t perf;
-      int algo_count = 1;
       int cudnn_conv_algo = cuda_ep->GetCudnnConvAlgo();
-      ORT_ENFORCE(cudnn_conv_algo > -1 && cudnn_conv_algo < 3, "cudnn_conv_algo should be 0, 1 or 2, but got ", cudnn_conv_algo);
+      cudnn_frontend::HeurMode_t heur_mode;
       switch (cudnn_conv_algo) {
-        case 0: {
-          static constexpr int num_algos = CUDNN_CONVOLUTION_FWD_ALGO_COUNT;
-          size_t max_ws_size = cuda_ep->GetCudnnConvUseMaxWorkspace() ? GetMaxWorkspaceSize(GetCudnnHandle(context), s_, kAllAlgos, num_algos)
-                                                                      : AlgoSearchWorkspaceSize;
-          // Use GetTransientScratchBuffer() so the workspace can be freed instead of cached.
-          // Because the benchmarking uses a huge amount of memory, e.g. a few GBs.
-          IAllocatorUniquePtr<void> algo_search_workspace = GetTransientScratchBuffer<void>(max_ws_size);
-          CUDNN_RETURN_IF_ERROR(cudnnFindConvolutionForwardAlgorithmEx(
-              GetCudnnHandle(context),
-              s_.x_tensor,
-              s_.x_data,
-              s_.w_desc,
-              s_.w_data,
-              s_.conv_desc,
-              s_.y_tensor,
-              s_.y_data,
-              1,            // requestedAlgoCount
-              &algo_count,  // returnedAlgoCount
-              &perf,
-              algo_search_workspace.get(),
-              max_ws_size));
+        case 0:
+          heur_mode = cudnn_frontend::HeurMode_t::B;
           break;
-        }
         case 1:
-          CUDNN_RETURN_IF_ERROR(cudnnGetConvolutionForwardAlgorithm_v7(
-              GetCudnnHandle(context),
-              s_.x_tensor,
-              s_.w_desc,
-              s_.conv_desc,
-              s_.y_tensor,
-              1,            // requestedAlgoCount
-              &algo_count,  // returnedAlgoCount
-              &perf));
+          heur_mode = cudnn_frontend::HeurMode_t::A;
           break;
-
         default:
-          perf.algo = kDefaultConvAlgo;
-          CUDNN_RETURN_IF_ERROR(GetWorkspaceSize(GetCudnnHandle(context), s_, perf.algo, &perf.memory));
-          if (std::is_same<T, MLFloat16>::value) {
-            perf.mathType = CUDNN_TENSOR_OP_MATH;
-          } else {
-            perf.mathType = CUDNN_DEFAULT_MATH;
+          heur_mode = cudnn_frontend::HeurMode_t::FALLBACK;
```

Comment:
Judging by the [cuDNN docs](https://docs.nvidia.com/deeplearning/cudnn/api/cudnn-graph-library.html?highlight=fallback#cudnnbackendheurmode-t) this does not guarantee performance. I would probably set `HEUR_A` as the default mode. Fallback should only be used if nothing was found.